### PR TITLE
Request to modify a notification payload.

### PR DESCRIPTION
**Pull request by acierto**
## Hey, I just made a Pull Request!

A Notification System is a really good base to start building on top of it.
One of the things which makes it hard to use at the moment is the fixed payload. 
Each company can have own vision what notification payload can be.
Therefore I propose to add an extra flexible JSON field, where the extra information can be stored. 
Postges can index JSON format well, so performance won't drop.

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: beps/0001-notifications-system/README.md
Comment (Rugvip) [inline beps/0001-notifications-system/README.md:151]: Yep I think we need this. It would be good to have a bit more clarity in who produces and consumes this though. Who owns the format of the extra fields and who is allowed to look at the data.

For example, fairly straightforward I think would be that a plugin that posts a notification can then use the extra information for customized rendering in the frontend. But are notification processors allowed to access that data for custom logic, and what are the stability guarantees of the data in that case?

**Review by Rugvip — COMMENTED** [review]
👍, bit of extra questioning that I think we need some high level answers around

File: beps/0001-notifications-system/README.md
Comment (acierto) [reply beps/0001-notifications-system/README.md:151]: We would like to build a template, where stakeholders from technical platform teams could notify all affected feature teams what has been changed and which applications were affected, so teams have to take an action, they have a deadline for that, etc. Once the action by team has been done they mark it as completed. This way we can get rid of using emails, which sometimes include also more teams then needed.

Most of the fields which are currently in "notification" table perfectly fit this purpose, but some fields are missing (of course 😄 ) which we would like to store.

In some mid-term we would like to add some other notifications which comes from security point of view, and the template, number of fields are different from the first template. 

Therefore having a "notification" for our case should be flexible to support it. 

Maybe we can have something like this:

```
{
  "data": {
    "deadline": "01-06-2024",
    "affectedTeams": [
      "team1",
      "team2"
    ]
  },
  "schema": {
    "module": "central-hub-notification-module",
    "fields": [
      {
        "name": "deadline",
        "format": date,
        "optional": false
      },
      {
        "affectedTeams": string[]
      }
    ]
  }
}
```

This can be an indication which module can process an 'extra' field. 

If to think more globally, even one more field in the "Notification" or "NotificationPayload" object could be added, **module owner/source owner**, and only that module can retrieve this notification. 

So it's a matter of limiting the access only to "extra" field, or completely filtering out notifications based on custom notification module

Comment (drodil) [reply beps/0001-notifications-system/README.md:151]: How about adding a `notification.metadata` as `JsonValue | undefined` to the notification payload and saving that as JSON to the database? Also, I think this data should be available for the processors to modify/read as needed. The default processors (email/slack/...) from upstream probably don't know what to do with this data and should not care but custom processors can do their things with the data.

Comment (drodil) [reply beps/0001-notifications-system/README.md:151]: Yeah, absolutely as part of the `notificationPayload`.

**Comment by drodil** [comment]
LGTM, are you willing to contribute a PR for the metadata once this is approved @acierto, or should @mareklibra or I do it at some point?

**Comment by acierto** [comment]
> LGTM, are you willing to contribute a PR for the metadata once this is approved @acierto, or should @mareklibra or I do it at some point?

Yeah I'm good to contribute it.
Waiting for an approval prior to start working on that 😄

File: beps/0001-notifications-system/README.md
Comment (Rugvip) [inline beps/0001-notifications-system/README.md:151]:
```suggestion
The metadata is an opaque JSON field, where an additional payload can be stored. The format of this data is owned by the notification sender and is tied to the notification topic, i.e. notifications sent from the source source on the same topic should use a compatible metadata format. The primary purpose of this field is to allow for custom processing and rendering based on the additional metadata.
```

Suggesting to be a bit more strict about the format of the metadata field. Most likely it's enough to document the format/type, but I could also see it being useful to even have an adjacent metadata version field, or even allow for multiple metadata buckets of different types, e.g.

```ts
  metadata: Array<{
    type: string;
    value: JsonValue;
  }>
```

Over all I'd like us to be very careful about making sure that this data is easy to consume in an explicit way, while making it possible to evolve the data over time. If we go forward with the multiple metadata idea I'd event suggest that we move some of the existing fields into metadata, like icon, links, etc.

**Review by Rugvip — CHANGES_REQUESTED** [review]

File: beps/0001-notifications-system/README.md
Comment (drodil) [reply beps/0001-notifications-system/README.md:151]: Additionally, we need some way for the other plugins to consume these notifications. Sure they can hook into the notifications signal but I am not sure if that's the proper way to distribute who wants which notification in the frontend.

Comment (acierto) [reply beps/0001-notifications-system/README.md:151]: @Rugvip How about to use `metadata` as `JsonObject` for multiple multiple metadata buckets?
The problem with the array in this model that querying it with knex is not really supported.

It still can be the same idea, where we can say something like:

```
metadata: {
  link: {
     type: 'string',
     value: 'https://github.com/backstage/backstage'
  },
  age: {
     type: 'number',
     value: 30
  }
}
```

And this i also don't like as a user who will work with this. 

The more convenient way might be to have 2 fields: one `metadata` and another `metadataSchema`.
Then it will look like

"metadata"
```
{
  link: 'https://github.com/backstage/backstage',
  age: : 30,
}
```

"metadataSchema"
```
{
  link: 'string',
  age: 'number',
}
```

Completely agree that "link", "icon", "etc." can be moved to "metadata".

Comment (acierto) [reply beps/0001-notifications-system/README.md:151]: @drodil Having a field "type" in notifications (could be an extra field in metadata as well) can help with filtering notifications per custom notification module.

Comment (Rugvip) [reply beps/0001-notifications-system/README.md]: @acierto PostgreSQL has excellent support for arrays - if you normalize the schema 😁, i.e. storing the metadata in a separate table.

With `type` I had something more strict in mind, things like `core.link`, `catalog.entity-ref`, `core.icon.v1`, etc.

Comment (acierto) [reply beps/0001-notifications-system/README.md]: Ah ok, that what you have in mind, I was thinking to use `jsonb` as a type of field and for that I couldn't find a solution to query arrays.

Comment (acierto) [reply beps/0001-notifications-system/README.md]: If to some up the idea:

We would like to have an extra table, i.e.
`notification_payalod_metadata` with columns:

type - string
value - jsonb
version - string

On code level side:

```
  metadata: Array<{
    type: string;
    value: JsonValue;
    version: string;
  }>
```

@Rugvip with having `type` more strict, do you mean that the types should have more context? 
As we can't/shouldn't validate the type content.

I see the last example as `core.icon.v1`, does that mean that you prefer to keep versioning in the type name instead of a separate field?

Comment (Rugvip) [reply beps/0001-notifications-system/README.md]: @acierto yes to all those things 😁

**Review by Rugvip — APPROVED** [review]
👍 to this update in its current form. I think we can leave the discussion for whether we should move more things to metadata for later. Curious what you think about that though @drodil, @mareklibra

**Comment by drodil** [comment]
I think this looks good after the good conversation about this. Feel free to merge!

**Review by mareklibra — APPROVED** [review]

**Pull request merged**
